### ISSUE TITLE
libcmetrics.a does not provide mpack and xxhash symbols

### ISSUE THREAD

**Issue by cosmo0920** [calyptia/cmetrics#6]
## TL;DR

Perhaps, we should do polish for install target work?
Or, xxhash and mpack libraries installation should be executed in the separated steps?

### Environment

* Ubuntu 18.04.5 LTS & macOS Big Sur 11.4
* I'm currently using forked version of cmetrics: https://github.com/calyptia/cmetrics/pull/4/commits/59f2f1e3dd9687bfc38e39caf0430445e3044acf

From llvm nm commands, libcmetrics.a only provides _cmt_xxx symbols:

```log
nm -gU ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a                               (git)[master]-[OK]

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_gauge.c.o):
0000000000000250 T _cmt_gauge_add
0000000000000000 T _cmt_gauge_create
0000000000000210 T _cmt_gauge_dec
0000000000000130 T _cmt_gauge_destroy
00000000000002f0 T _cmt_gauge_get_val
00000000000001d0 T _cmt_gauge_inc
0000000000000180 T _cmt_gauge_set
00000000000002a0 T _cmt_gauge_sub

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_counter.c.o):
00000000000001c0 T _cmt_counter_add
0000000000000000 T _cmt_counter_create
0000000000000130 T _cmt_counter_destroy
0000000000000270 T _cmt_counter_get_val
0000000000000180 T _cmt_counter_inc
0000000000000210 T _cmt_counter_set

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_metric.c.o):
00000000000000f0 T _cmt_metric_add
0000000000000080 T _cmt_metric_dec
00000000000001c0 T _cmt_metric_get_timestamp
00000000000001b0 T _cmt_metric_get_value
0000000000000030 T _cmt_metric_inc
0000000000000000 T _cmt_metric_set
0000000000000150 T _cmt_metric_sub

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_map.c.o):
0000000000000000 T _cmt_map_create
0000000000000200 T _cmt_map_destroy
0000000000000330 T _cmt_map_metric_get
0000000000000570 T _cmt_map_metric_get_val

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_log.c.o):
0000000000000110 T _cmt_errno_print
0000000000000000 T _cmt_log_print

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_opts.c.o):
00000000000004f0 T _cmt_opts_exit
0000000000000000 T _cmt_opts_init

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_time.c.o):
0000000000000000 T _cmt_time_now

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmetrics.c.o):
0000000000000000 T _cmt_create
0000000000000060 T _cmt_destroy

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_encode_prometheus.c.o):
0000000000000ca0 T _cmt_encode_prometheus_create
0000000000000eb0 T _cmt_encode_prometheus_destroy
00000000000000b0 T _format_metrics
0000000000000000 T _sds_cat_safe

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_encode_msgpack.c.o):
0000000000000290 T _cmt_encode_msgpack_to_cmetrics
0000000000000000 T _cmt_encode_msgpack_to_msgpack

ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_atomic_clang.c.o):
0000000000000010 T _cmt_atomic_compare_exchange
0000000000000000 T _cmt_atomic_initialize
0000000000000040 T _cmt_atomic_load
0000000000000030 T _cmt_atomic_store
```

mpack symbols and xxhash symbols are not provided:

```log
$ nm -g ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a
<snip>
ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_map.c.o):
                 U _XXH64_digest
                 U _XXH64_reset
                 U _XXH64_update
                 U ___error
                 U _calloc
                 U _cmt_errno_print
0000000000000000 T _cmt_map_create
0000000000000200 T _cmt_map_destroy
0000000000000330 T _cmt_map_metric_get
0000000000000570 T _cmt_map_metric_get_val
                 U _cmt_metric_get_value
                 U _free
                 U _malloc
                 U _memcpy
                 U _strlen
<snip>
ports/x86_64-apple-darwin20.5.0/cmetrics/master/lib/libcmetrics.a(cmt_encode_msgpack.c.o):
                 U ___stderrp
0000000000000290 T _cmt_encode_msgpack_to_cmetrics
0000000000000000 T _cmt_encode_msgpack_to_msgpack
                 U _cmt_metric_get_value
                 U _fwrite
                 U _mpack_start_array
                 U _mpack_start_map
                 U _mpack_write_cstr
                 U _mpack_write_double
                 U _mpack_write_u64
                 U _mpack_writer_destroy
                 U _mpack_writer_init_growable
```

I'm not sure these issues are intended.

This symbol absent issue causes the following (when using cmt_counter_create w/ labels):

```log
dyld: lazy symbol binding failed: Symbol not found: _XXH64_reset
  Referenced from: /Users/cosmo/GitHub/cmetrics-ruby/lib/cmetrics/cmetrics.bundle
  Expected in: flat namespace

dyld: Symbol not found: _XXH64_reset
  Referenced from: /Users/cosmo/GitHub/cmetrics-ruby/lib/cmetrics/cmetrics.bundle
  Expected in: flat namespace

zsh: abort      bundle exec irb
```

This repo provides testing examples for counter and gause, but they are linking xxhash.a and libmpack-static.a from lib directories:

### cmt-test-gause

cmetrics/test/CMakefiles/cmt-test-gauge.dir/link.txt contains:

```
/usr/bin/cc -O3 -DNDEBUG  -rdynamic CMakeFiles/cmt-test-gauge.dir/gauge.c.o  -o cmt-test-gauge ../src/libcmetrics.a ../lib/xxHash-0.8.0/cmake_unofficial/libxxhash.a ../lib/mpack/libmpack.a 
```

### cmt-test-counter

cmetrics/test/CMakefiles/cmt-test-counter.dir/link.txt contains:

```
/usr/bin/cc -O3 -DNDEBUG  -rdynamic CMakeFiles/cmt-test-counter.dir/counter.c.o  -o cmt-test-counter ../src/libcmetrics.a ../lib/xxHash-0.8.0/cmake_unofficial/libxxhash.a ../lib/mpack/libmpack-static.a 
```

**Additional context**

And I'm not confirmed yet but also it will causes symbol lookup issue on `_cmt_encode_msgpack_to_msgpack` or `_cmt_encode_msgpack_to_cmetrics`.

**Comment by edsiper** [comment]
looks like a linking issue... 

but for example if you create the tests -DCMT_TESTS=ON  the final binary contains the dependencies, so the caller must link the dependencies

**Comment by cosmo0920** [comment]
This should be fixed by compiling cmetrics as shared library or linking as 3rdparty library again like cool.io:
https://github.com/tarcieri/cool.io/blob/master/ext/cool.io/libev.c
